command/job_scaling_events.go

```diff
@@ -0,0 +1,189 @@
+func (j *JobScalingEventsCommand) Help() string {
+	helpText := `
+Usage: nomad job scaling-events [options] <args>
+
+  List the scaling events for the specified job.
+
+General Options:
+
+  ` + generalOptionsUsage() + `
+
+Scaling-Events Options:
+
+  -verbose
+    Display full information.
+`
+	return strings.TrimSpace(helpText)
+}
+
+// Synopsis satisfies the cli.Command Synopsis function.
+func (j *JobScalingEventsCommand) Synopsis() string {
+	return "Display the most recent scaling events for a job"
+}
+
+func (j *JobScalingEventsCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(j.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-verbose": complete.PredictNothing,
+		})
+}
+
+// Name returns the name of this command.
+func (j *JobScalingEventsCommand) Name() string { return "job scaling-events" }
+
+// Run satisfies the cli.Command Run function.
+func (j *JobScalingEventsCommand) Run(args []string) int {
+
+	var verbose bool
+
+	flags := j.Meta.FlagSet(j.Name(), FlagSetClient)
+	flags.Usage = func() { j.Ui.Output(j.Help()) }
+	flags.BoolVar(&verbose, "verbose", false, "")
+	if err := flags.Parse(args); err != nil {
+		return 1
+	}
+
+	args = flags.Args()
+	if len(args) != 1 {
+		j.Ui.Error("This command takes one argument: <job_id>")
+		j.Ui.Error(commandErrorText(j))
+		return 1
+	}
+
+	// Get the job ID.
+	jobID := args[0]
+
+	// Get the HTTP client.
+	client, err := j.Meta.Client()
+	if err != nil {
+		j.Ui.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+
+	events, _, err := client.Jobs().ScaleStatus(jobID, nil)
+	if err != nil {
+		j.Ui.Error(fmt.Sprintf("Error listing scaling events: %s", err))
+		return 1
+	}
+
+	// Check if any of the task groups have scaling events, otherwise exit
+	// indicating there are not any.
+	var haveEvents bool
+	for _, tg := range events.TaskGroups {
+		if tg.Events != nil {
+			haveEvents = true
+			break
+		}
+	}
+
+	if !haveEvents {
+		j.Ui.Output("No events found")
+		return 0
+	}
+
+	// Create our sorted list of events and output.
+	sortedList := sortedScalingEventList(events)
+	j.Ui.Output(formatList(formatScalingEventListOutput(sortedList, verbose, 0)))
+	return 0
+}
+
+func formatScalingEventListOutput(e scalingEventList, verbose bool, limit int) []string {
+
+	// If the limit is zero, aka no limit or the limit is greater than the
+	// number of events we have then set it this to the length of the event
+	// list.
+	if limit == 0 || limit > len(e) {
+		limit = len(e)
+	}
+
+	// Create the initial output heading.
+	output := make([]string, limit+1)
+	output[0] = "Task Group|Count|PrevCount"
+
+	// If we are outputting verbose information, add these fields to the header
+	// and then add our end date field.
+	if verbose {
+		output[0] += "|Error|Message|Eval ID"
+	}
+	output[0] += "|Date"
+
+	var i int
+
+	for i < limit {
+		output[i+1] = fmt.Sprintf("%s|%v|%v", e[i].name, *e[i].event.Count, e[i].event.PreviousCount)
```

Comment:
I got a `panic` here:

```shellsession
$ nomad job scaling-events webapp
No events found

$ nomad job scale webapp 10
Error submitting scaling request: Unexpected response code: 400 (job scaling blocked due to active deployment)

$ nomad job scaling-events webapp
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5d8a3f6]

goroutine 1 [running]:
github.com/hashicorp/nomad/command.formatScalingEventListOutput(0xc0001329c0, 0x1, 0x1, 0x0, 0x1, 0xc000656f20, 0x0, 0x0)
        github.com/hashicorp/nomad/command/job_scaling_events.go:135 +0x1b6
github.com/hashicorp/nomad/command.(*JobScalingEventsCommand).Run(0xc00025e580, 0xc000138030, 0x1, 0x1, 0xc000328560)
        github.com/hashicorp/nomad/command/job_scaling_events.go:108 +0x3ec
github.com/mitchellh/cli.(*CLI).Run(0xc0003ca640, 0xc0003ca640, 0xc0001286a0, 0x3c)
        github.com/mitchellh/cli@v1.1.0/cli.go:260 +0x1da
main.RunCustom(0xc000138010, 0x3, 0x3, 0xc00010e058)
        github.com/hashicorp/nomad/main.go:142 +0x482
main.Run(...)
        github.com/hashicorp/nomad/main.go:87
main.main()
        github.com/hashicorp/nomad/main.go:83 +0x64

```

When an event is an error `Count` is `nil`

```json
{
  "JobCreateIndex": 10,
  "JobID": "webapp",
  "JobModifyIndex": 23,
  "JobStopped": false,
  "Namespace": "default",
  "TaskGroups": {
    "demo": {
      "Desired": 3,
      "Events": [
        {
          "Count": null,
          "CreateIndex": 18,
          "Error": true,
          "EvalID": null,
          "Message": "job scaling blocked due to active deployment",
          "Meta": {
            "OriginalMessage": "submitted using the Nomad CLI",
            "OriginalCount": 10,
            "OriginalMeta": null
          },
          "PreviousCount": 3,
          "Time": 1601928117680287000
        }
      ],
      "Healthy": 3,
      "Placed": 3,
      "Running": 3,
      "Unhealthy": 0
    }
  }
}
```

A column to indicate if the event is an error could also be helpful.

Reply:
Nice find! This also impacts the verbose flag with the `EvalID` so will fix this for both areas.